rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp

```diff
@@ -276,6 +276,68 @@ TYPED_TEST(TestAddCallbackGroupsToExecutor, one_node_many_callback_groups_many_e
   ASSERT_EQ(timer_executor.get_all_callback_groups().size(), 2u);
 }
 
+/*
+ * Test callback groups from one node to many executors.
+ * A subscriber on a new executor with a callback group not received a message
+ * because the executor can't be triggered while a subscriber created, see
+ * https://github.com/ros2/rclcpp/issues/1611
+*/
+TYPED_TEST(TestAddCallbackGroupsToExecutor, subscriber_triggered_to_receive_message)
+{
+  auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+
+  // create a thread running an executor with a new callback group for a coming subscriber
+  rclcpp::CallbackGroup::SharedPtr cb_grp = node->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive, false);
+  rclcpp::executors::SingleThreadedExecutor cb_grp_executor;
+  std::thread cb_grp_thread = std::thread(
+    [&cb_grp, &node, &cb_grp_executor]() {
+      cb_grp_executor.add_callback_group(cb_grp, node->get_node_base_interface());
+      cb_grp_executor.spin();
+    });
+
+  // expect the subscriber to receive a message
+  bool received_message = false;
+  auto sub_callback = [&received_message](const test_msgs::msg::Empty::SharedPtr) {
+      received_message = true;
+    };
+
+  // a timer callback run on the default executor
+  rclcpp::TimerBase::SharedPtr timer = nullptr;
+  auto timer_callback = [&timer, &cb_grp, &node, &sub_callback]() {
+      if (timer) {
+        timer.reset();
+      }
+
+      // create a subscription using the `cb_grp` callback group
+      rclcpp::QoS qos = rclcpp::QoS(1).reliable();
+      auto options = rclcpp::SubscriptionOptions();
+      options.callback_group = cb_grp;
+      auto subscription =
+        node->create_subscription<test_msgs::msg::Empty>("topic_name", qos, sub_callback, options);
+
+      // create a publisher to send data
+      auto publisher =
+        node->create_publisher<test_msgs::msg::Empty>("topic_name", qos);
+      publisher->publish(test_msgs::msg::Empty());
+      // wait for message to be published
+      std::this_thread::sleep_for(100ms);
```

Comment:
When `subscription` and `publisher` go out of scope they will be removed from the graph, and the 100ms sleep is unlikely to be enough all the time, i.e. this will be flaky.

Is it important to the test that these be in the scope of the timer callback? If not, then move them out, if so, then maybe this isn't an issue with the middleware, but how you're using the API.

Reply:
> Is it important to the test that these be in the scope of the timer callback?

I want to use the simplest source code to emulate the reported example of this issue.
timer callback ---> service callback
create a subscription on the service callback
there is a remote publisher to publish a message to the subscription,
that's all to reproduce this issue.

Reply:
`100ms` might be caused this test case to be flaky, I'll update it.

Reply:
> I want to use the simplest source code to emulate the reported example of this issue.

Yes, but do you understand that not keeping these in scope might cause issues? I don't think (I could be wrong) that is what the original problem was surrounding.

Also, sleeping in the timer callback will block anything else (either something you set up or functions of the node that are implementation details) from happening via callback (for a single threaded executor) so you should never do that if you can avoid it.

Reply:
Please refer to https://github.com/ros2/rclcpp/issues/1611#issue-846709308 reported by user,

```
  void srv_callback(const std_srvs::srv::Trigger::Request::SharedPtr, std_srvs::srv::Trigger::Response::SharedPtr res){
    RCLCPP_INFO_STREAM(this->get_logger(), "Received request");
    new_msg_ = false;

    // Create subscription
    auto sub_options_ = rclcpp::SubscriptionOptions();
    sub_options_.callback_group = callback_group_;
    auto subscription = this->create_subscription<std_msgs::msg::Bool>(
          "topic_bool",
          rclcpp::QoS(1),
          std::bind(
            &CallbackGroupSrvServer::sub_callback,
            this,
            std::placeholders::_1),
          sub_options_
          );

    // Wait for new_msg
    while (rclcpp::ok()
           && new_msg_ == false) {
      rclcpp::sleep_for(std::chrono::seconds(1));
      RCLCPP_INFO_STREAM(this->get_logger(), "Sleep...");
    }

    res->message = "";
    res->success = true;
    return;
  }
```

`srv_callback` running on the main thread, and the `CallbackGroupSrvServer::sub_callback` running on the `callback_group_executor_thread_` thread, so it seems it's a normal case. (publishing message can run on the main thread), so I think it's not a problem.

Reply:
And the user said `which we use a lot.`

Reply:
I think that we'd better use the same case (but using three applications is unnecessary) to reproduce the issue by adding a regression test,

Reply:
I don't want to argue about whether it is a good design or not. I mean It's just a use case somebody might use.

Reply:
> I don't want to argue about whether it is a good design or not. I mean It's just a use case somebody might use.

Yes, but it is an anti-pattern (two of them):

- not keeping the pub/sub in scope and assuming some short period of time is long enough for the message to arrive
- blocking in a service/timer (or anything) callback

We could warn or prevent these if it were possible, but we cannot prevent the user from doing so or know when to warn and when not to.

This is a test, not some random user code, so we should avoid these issues. And doing so we should figure out what about the user reported code is the source of the problem. Is it the blocking in the callback? Is it the pub/sub going out of scope? Is it a bug in rclcpp? It's hard to tell without eliminating the obviously bad things to do.